flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java

```diff
@@ -3994,6 +4008,130 @@ public void testSharedStateRegistrationOnRestore() throws Exception {
 		}
 	}
 
+	@Test
+	public void jobFailsIfInFlightSynchronousSavepointIsDiscarded() throws Exception {
+		final JobID jobId = new JobID();
+
+		final ExecutionGraph executionGraph = getExecutionGraph(jobId);
+		final CheckpointCoordinator coordinator = enableCheckpointingAndGetCoordinator(executionGraph);
+		final ExecutionAttemptID attemptID = executeAndGetAttemptID(executionGraph);
+
+		assertEquals(JobStatus.RUNNING, executionGraph.getState());
+
+		final CompletableFuture<CompletedCheckpoint> savepointFuture = coordinator
+				.triggerSynchronousSavepoint(10L, false, "test-dir");
+
+		final Throwable expectedRootCause = new IOException("Custom-Exception");
+		final PendingCheckpoint syncSavepoint = declineSynchronousSavepoint(jobId, coordinator, attemptID, expectedRootCause);
+		assertTrue(syncSavepoint.isDiscarded());
+
+		final JobStatus status = executionGraph.getState();
+		coordinator.shutdown(status);
+
+		assertEquals(JobStatus.FAILING, status);
+
+		final AtomicReference<Throwable> actualCause = new AtomicReference<>();
+		savepointFuture.handle((path, throwable) -> {
+			if (throwable != null) {
+				actualCause.set(throwable);
+			}
+			return path;
+		}).get();
+
+		assertThat(actualCause.get(), equals(expectedRootCause));
```

Comment:
Consider `ExceptionUtils#stripCompletionException()` and `ExceptionUtils#findThrowableWithMessage()`